File: backend/modules/market/providers/jupiter.py
"""Jupiter adapter — price cross-check from the DEX aggregator's view."""

from modules.market.market_models import ProviderQuote
from modules.market.providers import MarketProvider


class JupiterProvider(MarketProvider):
    """https://lite-api.jup.ag — free lite tier, no key required."""

    name = "jupiter"

    async def _fetch(self, mint: str) -> ProviderQuote | None:
        res = await self._http.get(f"https://lite-api.jup.ag/price/v3?ids={mint}")
        res.raise_for_status()
        d = res.json().get(mint)
        if not d or d.get("usdPrice") is None:
            return None
        return ProviderQuote(
            provider=self.name,
            price_usd=float(d["usdPrice"]),
            change_24h=d.get("priceChange24h"),
            liquidity_usd=d.get("liquidity"),
        )

File: 09 Automation/market/sources/jupiter.mjs
// Jupiter — price cross-check + swap routing (free lite tier, no key)
export async function jupPrices(mints) {
  const res = await fetch(
    `https://lite-api.jup.ag/price/v3?ids=${mints.join(",")}`,
  );
  if (!res.ok) throw new Error(`Jupiter HTTP ${res.status}`);
  const json = await res.json();
  const out = {};
  for (const [mint, d] of Object.entries(json)) {
    if (d?.usdPrice != null) out[mint] = Number(d.usdPrice);
  }
  return out; // { mint: priceUsd }
}

// Swap route quote — used by paper trading to sanity-check executability.
export async function jupQuote(inputMint, outputMint, amountLamports) {
  const res = await fetch(
    `https://lite-api.jup.ag/swap/v1/quote?inputMint=${inputMint}&outputMint=${outputMint}&amount=${amountLamports}&slippageBps=50`,
  );
  if (!res.ok) throw new Error(`Jupiter quote HTTP ${res.status}`);
  return await res.json();
}
